Encapsulate FASTA/Q parsing, ignore reads with wildcards (for now)

tsnorri opened the pull request:
* Added an interface for fetching the reads from the input, `reader_adapter`.
* Currently the FASTA/Q parser in use is the one from SeqIO, since libbio’s parser did not seem to affect performance in any way. I think we should switch, though, since the latter validates the input. The downside is that [Ragel](https://www.colm.net/open-source/ragel/) is needed to build libbio.
* Wildcards are now handled by discarding the reads per Kimmo’s suggestion (please see #23). I’ll add the functionality to make use of the reads where possible as soon as I can.
* Names of the input files with skipped reads are also printed to stderr. When libbio’s parsers are used, the line numbers are also printed.
* gapmers are now constructed from packed characters. I made this change mostly because libbio’s parsers pass `std::string_view`s instead of copying the input and hence packing the characters while parsing made sense. I decided to keep it anyway since gapmer’s constructors expect valid inputs. 
* This also fixes the bug which caused wildcard characters to be replaced with various characters depending on the situation including A and G.
* Some header-only functionality from libbio is now used so added it as a dependency.
* Applied conventional variable names to Makefile.

tsnorri committed:
Add next_()

– The functions take 2-bit encoded characters as parameters.

tsnorri committed:
Add libbio as submodule, use FASTA, FASTQ parsers as part of gapmer_count, partial_count

tsnorri committed:
Ignore unused function warnings

tsnorri committed:
Update libbio

tsnorri committed:
Ignore user-specific files, build products

tsnorri committed:
Fix next_, add static variable for middle_gap_only

tsnorri committed:
Move packed character iteration to its own header, add tests

tsnorri committed:
Fix template function call

– Clang reported this as error.

tsnorri committed:
Fix handling gap start, report reads with unexpected characters

tsnorri committed:
Remove old comment

tsnorri committed:
Fix test headers

tsnorri committed:
Add constructor for packed character buffer, tests

tsnorri committed:
Make use of gapmer’s new constructor

tsnorri committed:
Make pack_characters visible in namespace sf, add test

tsnorri committed:
Add seqio_reader_adapter, common reader adapter delegate interface, fix character iteration in partial_count, fix error reporting in libbio_reader_adapter

– SeqIO’s FASTA/Q reader is used again for parsing the input.
– A quick benchmark indicated that using libbio’s FASTQ reader does not improve performance. It does validate the input, though, so it may be worthwhile to use it anyway.
– partial_count now refrains from comparing the read length to the value of k (see #22). The check should probably be done, though.
– partial_count now iterates the character pairs in the input reads starting from the correct index.

tsnorri committed:
Merge remote-tracking branch 'saskeli/main' into improvement/libbio-fastx-reader

tsnorri committed:
Fix building tests

tsnorri committed:
Apply IWYU

tsnorri committed:
Refrain from building libbio.a

– Only the headers are used currently.

tsnorri committed:
Update targets, make use of conventional variable names, re-add all target

– The “all” target got removed at some point which caused “motivating” to be the default target. This is now fixed.
– Added variables that follow the well-established naming convention.

tsnorri committed:
Format code

tsnorri committed:
Remove libboost_iostreams from libraries

– The library is needed by libbio so it need not be linked at the moment.

saskeli reviewed (commented):
There's way too much stuff with way too many contexts to effectively review here.

I feel like we wouldn't necessarily need boost, ragel and libbio just to get fasta validation, that I'm not sure we need? 

Also `error: ‘-Wno-error=unused-command-line-argument’: no option ‘-Wunused-command-line-argument’`

Also, also, the dot_writer is dropped from the Makefile.

tsnorri committed:
Document from_packed_characters(std::span<uint64_t const>, uint8_t, uint8_t, uint8_t), improve comment formatting

tsnorri committed:
Format

tsnorri committed:
Comment from_packed_characters

tsnorri committed:
Rename from_packed_characters to from_value when semantics are different

– The member functions in question took packed characters such that the characters in the gap range were either included or excluded. Changed the function names in such a way that the different semantics are clearer.

tsnorri committed:
Document reader_adapter

tsnorri committed:
Improve comments

tsnorri committed:
Add comments to FromPackedCharacterStringConstructor, format

tsnorri committed:
Add comments

tsnorri committed:
Improve comment, formatting

tsnorri committed:
Format code

tsnorri committed:
Fix function call

tsnorri committed:
Add sanity check

tsnorri committed:
Fix vector initialisation

– sf::pack_characters appends to the passed vector, so we should not allocate its contents like before.
